// @elastic/elasticsearch@9.2.0/lib/api/api/synonyms.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and contributors
 * SPDX-License-Identifier: Apache-2.0
 */
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
const client_1 = require("../../client");
const commonQueryParams = ['error_trace', 'filter_path', 'human', 'pretty'];
class Synonyms {
    constructor(transport) {
        Object.defineProperty(this, "transport", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, _a, {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.transport = transport;
        this[client_1.kAcceptedParams] = {
            'synonyms.delete_synonym': {
                path: [
                    'id'
                ],
                body: [],
                query: []
            },
            'synonyms.delete_synonym_rule': {
                path: [
                    'set_id',
                    'rule_id'
                ],
                body: [],
                query: [
                    'refresh'
                ]
            },
            'synonyms.get_synonym': {
                path: [
                    'id'
                ],
                body: [],
                query: [
                    'from',
                    'size'
                ]
            },
            'synonyms.get_synonym_rule': {
                path: [
                    'set_id',
                    'rule_id'
                ],
                body: [],
                query: []
            },
            'synonyms.get_synonyms_sets': {
                path: [],
                body: [],
                query: [
                    'from',
                    'size'
                ]
            },
            'synonyms.put_synonym': {
                path: [
                    'id'
                ],
                body: [
                    'synonyms_set'
                ],
                query: [
                    'refresh'
                ]
            },
            'synonyms.put_synonym_rule': {
                path: [
                    'set_id',
                    'rule_id'
                ],
                body: [
                    'synonyms'
                ],
                query: [
                    'refresh'
                ]
            }
        };
    }
    async deleteSynonym(params, options) {
        const { path: acceptedPath } = this[client_1.kAcceptedParams]['synonyms.delete_synonym'];
        const userQuery = params === null || params === void 0 ? void 0 : params.querystring;
        const querystring = userQuery != null ? { ...userQuery } : {};
        let body;
        const userBody = params === null || params === void 0 ? void 0 : params.body;
        if (userBody != null) {
            if (typeof userBody === 'string') {
                body = userBody;
            }
            else {
                body = { ...userBody };
            }
        }
        for (const key in params) {
            if (acceptedPath.includes(key)) {
                continue;
            }
            else if (key !== 'body' && key !== 'querystring') {
                // @ts-expect-error
                querystring[key] = params[key];
            }
        }
        const method = 'DELETE';
        const path = `/_synonyms/${encodeURIComponent(params.id.toString())}`;
        const meta = {
            name: 'synonyms.delete_synonym',
            pathParts: {
                id: params.id
            },
            acceptedParams: [
                'id'
            ]
        };
        return await this.transport.request({ path, method, querystring, body, meta }, options);
    }
    async deleteSynonymRule(params, options) {
        const { path: acceptedPath } = this[client_1.kAcceptedParams]['synonyms.delete_synonym_rule'];
        const userQuery = params === null || params === void 0 ? void 0 : params.querystring;
        const querystring = userQuery != null ? { ...userQuery } : {};
        let body;
        const userBody = params === null || params === void 0 ? void 0 : params.body;
        if (userBody != null) {
            if (typeof userBody === 'string') {
                body = userBody;
            }
            else {
                body = { ...userBody };
            }
        }
        for (const key in params) {
            if (acceptedPath.includes(key)) {
                continue;
            }
            else if (key !== 'body' && key !== 'querystring') {
                // @ts-expect-error
                querystring[key] = params[key];
            }
        }
        const method = 'DELETE';
        const path = `/_synonyms/${encodeURIComponent(params.set_id.toString())}/${encodeURIComponent(params.rule_id.toString())}`;
        const meta = {
            name: 'synonyms.delete_synonym_rule',
            pathParts: {
                set_id: params.set_id,
                rule_id: params.rule_id
            },
            acceptedParams: [
                'set_id',
                'rule_id',
                'refresh'
            ]
        };
        return await this.transport.request({ path, method, querystring, body, meta }, options);
    }
    async getSynonym(params, options) {
        const { path: acceptedPath } = this[client_1.kAcceptedParams]['synonyms.get_synonym'];
        const userQuery = params === null || params === void 0 ? void 0 : params.querystring;
        const querystring = userQuery != null ? { ...userQuery } : {};
        let body;
        const userBody = params === null || params === void 0 ? void 0 : params.body;
        if (userBody != null) {
            if (typeof userBody === 'string') {
                body = userBody;
            }
            else {
                body = { ...userBody };
            }
        }
        for (const key in params) {
            if (acceptedPath.includes(key)) {
                continue;
            }
            else if (key !== 'body' && key !== 'querystring') {
                // @ts-expect-error
                querystring[key] = params[key];
            }
        }
        const method = 'GET';
        const path = `/_synonyms/${encodeURIComponent(params.id.toString())}`;
        const meta = {
            name: 'synonyms.get_synonym',
            pathParts: {
                id: params.id
            },
            acceptedParams: [
                'id',
                'from',
                'size'
            ]
        };
        return await this.transport.request({ path, method, querystring, body, meta }, options);
    }
    async getSynonymRule(params, options) {
        const { path: acceptedPath } = this[client_1.kAcceptedParams]['synonyms.get_synonym_rule'];
        const userQuery = params === null || params === void 0 ? void 0 : params.querystring;
        const querystring = userQuery != null ? { ...userQuery } : {};
        let body;
        const userBody = params === null || params === void 0 ? void 0 : params.body;
        if (userBody != null) {
            if (typeof userBody === 'string') {
                body = userBody;
            }
            else {
                body = { ...userBody };
            }
        }
        for (const key in params) {
            if (acceptedPath.includes(key)) {
                continue;
            }
            else if (key !== 'body' && key !== 'querystring') {
                // @ts-expect-error
                querystring[key] = params[key];
            }
        }
        const method = 'GET';
        const path = `/_synonyms/${encodeURIComponent(params.set_id.toString())}/${encodeURIComponent(params.rule_id.toString())}`;
        const meta = {
            name: 'synonyms.get_synonym_rule',
            pathParts: {
                set_id: params.set_id,
                rule_id: params.rule_id
            },
            acceptedParams: [
                'set_id',
                'rule_id'
            ]
        };
        return await this.transport.request({ path, method, querystring, body, meta }, options);
    }
    async getSynonymsSets(params, options) {
        const { path: acceptedPath } = this[client_1.kAcceptedParams]['synonyms.get_synonyms_sets'];
        const userQuery = params === null || params === void 0 ? void 0 : params.querystring;
        const querystring = userQuery != null ? { ...userQuery } : {};
        let body;
        const userBody = params === null || params === void 0 ? void 0 : params.body;
        if (userBody != null) {
            if (typeof userBody === 'string') {
                body = userBody;
            }
            else {
                body = { ...userBody };
            }
        }
        params = params !== null && params !== void 0 ? params : {};
        for (const key in params) {
            if (acceptedPath.includes(key)) {
                continue;
            }
            else if (key !== 'body' && key !== 'querystring') {
                // @ts-expect-error
                querystring[key] = params[key];
            }
        }
        const method = 'GET';
        const path = '/_synonyms';
        const meta = {
            name: 'synonyms.get_synonyms_sets',
            acceptedParams: [
                'from',
                'size'
            ]
        };
        return await this.transport.request({ path, method, querystring, body, meta }, options);
    }
    async putSynonym(params, options) {
        const { path: acceptedPath, body: acceptedBody, query: acceptedQuery } = this[client_1.kAcceptedParams]['synonyms.put_synonym'];
        const userQuery = params === null || params === void 0 ? void 0 : params.querystring;
        const querystring = userQuery != null ? { ...userQuery } : {};
        let body;
        const userBody = params === null || params === void 0 ? void 0 : params.body;
        if (userBody != null) {
            if (typeof userBody === 'string') {
                body = userBody;
            }
            else {
                body = { ...userBody };
            }
        }
        for (const key in params) {
            if (acceptedBody.includes(key)) {
                body = body !== null && body !== void 0 ? body : {};
                // @ts-expect-error
                body[key] = params[key];
            }
            else if (acceptedPath.includes(key)) {
                continue;
            }
            else if (key !== 'body' && key !== 'querystring') {
                if (acceptedQuery.includes(key) || commonQueryParams.includes(key)) {
                    // @ts-expect-error
                    querystring[key] = params[key];
                }
                else {
                    body = body !== null && body !== void 0 ? body : {};
                    // @ts-expect-error
                    body[key] = params[key];
                }
            }
        }
        const method = 'PUT';
        const path = `/_synonyms/${encodeURIComponent(params.id.toString())}`;
        const meta = {
            name: 'synonyms.put_synonym',
            pathParts: {
                id: params.id
            },
            acceptedParams: [
                'id',
                'synonyms_set',
                'refresh'
            ]
        };
        return await this.transport.request({ path, method, querystring, body, meta }, options);
    }
    async putSynonymRule(params, options) {
        const { path: acceptedPath, body: acceptedBody, query: acceptedQuery } = this[client_1.kAcceptedParams]['synonyms.put_synonym_rule'];
        const userQuery = params === null || params === void 0 ? void 0 : params.querystring;
        const querystring = userQuery != null ? { ...userQuery } : {};
        let body;
        const userBody = params === null || params === void 0 ? void 0 : params.body;
        if (userBody != null) {
            if (typeof userBody === 'string') {
                body = userBody;
            }
            else {
                body = { ...userBody };
            }
        }
        for (const key in params) {
            if (acceptedBody.includes(key)) {
                body = body !== null && body !== void 0 ? body : {};
                // @ts-expect-error
                body[key] = params[key];
            }
            else if (acceptedPath.includes(key)) {
                continue;
            }
            else if (key !== 'body' && key !== 'querystring') {
                if (acceptedQuery.includes(key) || commonQueryParams.includes(key)) {
                    // @ts-expect-error
                    querystring[key] = params[key];
                }
                else {
                    body = body !== null && body !== void 0 ? body : {};
                    // @ts-expect-error
                    body[key] = params[key];
                }
            }
        }
        const method = 'PUT';
        const path = `/_synonyms/${encodeURIComponent(params.set_id.toString())}/${encodeURIComponent(params.rule_id.toString())}`;
        const meta = {
            name: 'synonyms.put_synonym_rule',
            pathParts: {
                set_id: params.set_id,
                rule_id: params.rule_id
            },
            acceptedParams: [
                'set_id',
                'rule_id',
                'synonyms',
                'refresh'
            ]
        };
        return await this.transport.request({ path, method, querystring, body, meta }, options);
    }
}
_a = client_1.kAcceptedParams;
exports.default = Synonyms;
